shakenfist/daemons/queues.py

```diff
@@ -0,0 +1,182 @@
+
+            if task.get('type').startswith('instance_'):
+                instance_uuid = workitem.get('instance_uuid')
+                db.add_event('instance', instance_uuid,
+                             task.get('type').replace('_', ' '), 'dequeued', None, None)
+
+            if task.get('type') == 'image_fetch':
+                image_fetch(task.get('url'))
+
+            if task.get('type') == 'instance_preflight':
+                instance_preflight(instance_uuid)
+
+            if task.get('type') == 'instance_start':
+                instance_start(instance_uuid, workitem.get('network'))
+                db.update_instance_state(instance_uuid, 'created')
+
+            if task.get('type') == 'instance_delete':
+                instance_delete(instance_uuid)
+                db.update_instance_state(instance_uuid,
+                                         task.get('next_state', 'unknown'))
+
+        LOG.info('Worker for workitem %s has pid %d, complete'
+                 % (jobname, os.getpid()))
+
+    finally:
+        db.resolve(config.parsed.get('NODE_NAME'), jobname)
+        if instance_uuid:
+            db.add_event('instance', instance_uuid,
+                         'tasks complete', 'dequeued', None, None)
+
+
+def image_fetch(url):
+    try:
+        images.get_image(url, [], url, timeout=images.IMAGE_FETCH_LOCK_TIMEOUT)
+    except etcd.LockException:
+        pass
+
+
+def instance_preflight(instance_uuid):
+    # TODO(mikal): preflight with retries etc
+    pass
+
+
+def instance_start(instance_uuid, network):
+    with db.get_lock('instance', None, instance_uuid, ttl=900) as lock:
+        db.add_event('instance', instance_uuid,
+                     'queued', 'create', None, None)
+        instance = virt.from_db(instance_uuid)
+
+        # Collect the networks
+        nets = {}
+        for netdesc in network:
+            if netdesc['network_uuid'] not in nets:
+                n = net.from_db(netdesc['network_uuid'])
+                if not n:
+                    db.enqueue_delete(
+                        config.parsed.get('NODE_NAME'), instance_uuid, 'error')
+                    return
+
+                nets[netdesc['network_uuid']] = n
+
+        # Create the networks
+        with util.RecordedOperation('ensure networks exist', instance) as _:
+            for network_uuid in nets:
+                n = nets[network_uuid]
+                n.create()
+                n.ensure_mesh()
+                n.update_dhcp()
+
+        # Now we can start the isntance
+        libvirt = util.get_libvirt()
+        try:
+            with util.RecordedOperation('instance creation',
+                                        instance) as _:
+                instance.create(lock=lock)
+
+        except libvirt.libvirtError as e:
+            code = e.get_error_code()
+            if code in (libvirt.VIR_ERR_CONFIG_UNSUPPORTED,
+                        libvirt.VIR_ERR_XML_ERROR):
+                db.enqueue_delete(
```

Comment:
I was thinking to log the error code in the log. It can be frustrating to not know the exact error especially when the software did know but didn't share.